### PR TITLE
VQ-GPT + amp (2x speedup) + new record 86.5% val accuracy with 64 tok bottleneck

### PR DESCRIPTION
**Pull request by iejMac**

### CODE REVIEW

**Comment by iejMac** [comment]
TODOs:
- [x] Try new AR approach (f0, f1, ..., xn-1, fn-1) -> xn
- [ ] ....maybe also just give it access to x<n-1...
- [x] In encoder try taking from above x1 not x0

**Comment by iejMac** [comment]
More context should make you better so later predictions should be better. Debug TODO's:
- [x] Make sure you are selecting correct logits
- [x] Try to set up the previous case but with many samples per sequence (all should develop the same way)
- [ ] Using the diff codes should actually improve results...
- [x] Fix encoder dynamic_inds

**Comment by iejMac** [comment]
I should set up a bunch of small experiments to verify why things aren't working: 
- [ ] If you only compute loss based on last frame then it should be super good because it has a much wider bottleneck (n_frames per frame prediction)
- [ ] If you use n_frames == 2 then you get same results
- [ ] Maybe try stronger decoder? Allow n-shot frame prediction? (128 shot is full AR token by token)
- [ ] Do small model experiments (<7M params, 3 or 4 frames)
- [ ] Final Goal: AR where predicting xn from x0-xn-1 > x1 from x0
